### PR TITLE
API Support

### PR DESCRIPTION
**Pull request by gokmen**
## Description
This PR will provide REST api for Koding Node backend with proper cookie or token or for public requests. **This is currently experimental and wip.**

## Screenshots (if appropriate):

Full screencast: http://recordit.co/AvGdOFRfu6

## TODO
 - [x] Add support for descriptions in schema on `Bongo::Model`
 - [x] Generate initial model definitions for Swagger api generator from Bongo models
 - [x] Generate api endpoints for static methods of models
 - [x] Generate api endpoints for instance methods of models
 - [x] Remote API handler for initial requests
 - [x] Enable throttling on remote endpoint
 - [x] Setup Swagger-UI in place to provide auto generated api docs
 - [x] Auto generate swagger.yaml on model change, serve it under website
 - [x] Provide a way to login on apidocs ui and embed `clientId` into it
 - [x] Support static methods of shared models
 - [x] Support instance methods of shared models
 - [x] Use existing `clientId` from cookies as authorization token
 - [x] Fix unhandled Bongo calls 
 - [x] Fail when invalid signatures provided
 - [x] Provide possible valid signatures with validation error
 - [x] Provide search functionality in api docs viewer
 - [x] Generate parameter types, examples and description from docs
 - [x] Compare source code and swagger.json against changes
  _example failed build https://app.wercker.com/buildstep/58346a4b62c4510100c29b8a if it has outdated swagger.json_
  - [ ] Tests needs to be written for the api cc/ @usirin

It has now embed documents from source code feature in this PR uses Codo library https://github.com/coffeedoc/codo we will eventually use this to generate human readable code documentation as well.

image

### CODE REVIEW

File: node_modules_koding/bongo/lib/bongo.coffee [hunk 1]
```diff
@@ -495,62 +499,84 @@ module.exports = class Bongo extends EventEmitter
...
+        if (kallback = @localStore.get method)?
           apply this, null, kallback, args, sessionToken, userArea, client
         else
-          @handleUserError 'Unknown method type!'
+          @handleUserError 'Unknown method type!', args
+
+
+  handleUserError: (message, args) ->
+
+    console.warn '[Bongo:UserError]', message, args
```
Comment (cihangir) [inline node_modules_koding/bongo/lib/bongo.coffee]: 👍

File: node_modules_koding/bongo/lib/expressify.coffee [hunk 2]
```diff
@@ -134,13 +138,20 @@ module.exports = (options = {}) ->
             return res.status(500).send 'An error occcurred'
 
           payload = []
+          healthyQueue = yes
 
           queue.forEach (_req) ->
-            if _req.method is 'authenticateUser'
+            unless _req.callbacks
+              healthyQueue = no
+              return res.status(500).send 'An error occcurred'
```
Comment (cihangir) [inline node_modules_koding/bongo/lib/expressify.coffee]: what about a more informative message or a log statement?

File: workers/social/lib/social/models/account.coffee [hunk 3]
```diff
@@ -56,28 +56,14 @@ module.exports = class JAccount extends jraphical.Module
         ]
         some:
           (signature Object, Object, Function)
-        cursor:
```
Comment (cihangir) [inline workers/social/lib/social/models/account.coffee]: 👍

**Review by cihangir — APPROVED** [review]

**Comment by cihangir** [comment]
could you please make sure swagger.json is always up-to-date with the current codebase in CI env?

File: workers/social/lib/social/models/session.coffee [hunk 4]
```diff
@@ -57,6 +61,10 @@ module.exports = class JSession extends Model
           console.error err  if err?
 
 
+  @activeSession = secure (client, callback) ->
+    @one { clientId: client.sessionToken }, callback
```
Comment (szkl) [inline workers/social/lib/social/models/session.coffee]: Would it be good to match username and group too?

Comment (gokmen) [inline workers/social/lib/social/models/session.coffee]: I don't have those information, just getting token from header. But using the `username` and `group` in returned `JSession` object.

Comment (szkl) [inline workers/social/lib/social/models/session.coffee]: Okay. I thought `client` would have them.

**Comment by gokmen** [comment]
> could you please make sure swagger.json is always up-to-date with the current codebase in CI env?

@cihangir sure thing! At first I was planning to not to keep swagger.json in repo but then I thought it might be a good thing to see what is changed on api. I'll add a wercker step to verify swagger.json with the code.

File: workers/social/lib/social/models/session.coffee (hunk 4 above)
Comment (gokmen) [inline workers/social/lib/social/models/session.coffee]: ah sorry, I thought you were asking for some place else, you're right I can add that check here. 👊

**Comment by cihangir** [comment]
I also liked the idea of keeping swagger def in the repo 👍  💯

**Comment by gokmen** [comment]
I'm done with this shall we merge? cc/ @cihangir @usirin @szkl @rjeczalik

**Review by rjeczalik — APPROVED** [review]

File: workers/social/lib/social/remoteapi.coffee [hunk 5]
```diff
@@ -0,0 +1,188 @@
+KONFIG    = require 'koding-config-manager'
```
Comment (cihangir) [inline workers/social/lib/social/remoteapi.coffee]: could you write test for this file?

File: workers/social/lib/social/models/session.coffee [hunk 6]
```diff
@@ -57,6 +61,18 @@ module.exports = class JSession extends Model
...
 
+  @activeSession = secure (client, callback) ->
+
+    username = client.connection?.delegate?.profile?.nickname
+    clientId = client.sessionToken
+
+    if not username or not clientId
+      return callback new KodingError 'Access denied'
+
+    @one { username, clientId }, callback
```
Comment (cihangir) [inline workers/social/lib/social/models/session.coffee]: clientId is enough, we dont need username + we dont have index on these two

File: workers/social/lib/social/models/computeproviders/providerinterface.coffee [hunk 7]
```diff
@@ -11,8 +11,17 @@ NOT_IMPLEMENTED = ->
 
   return NOT_IMPLEMENTED_MESSAGE
 
-PASS_THROUGH = (..., callback) -> callback null
-
+PASS_THROUGH = (rest... , callback) -> callback null
+
+# Base class for all providers
+#
+# @example How to subclass a provider
```
Comment (cihangir) [inline workers/social/lib/social/models/computeproviders/providerinterface.coffee]: 👍

File: wercker.yml [hunk 8]
```diff
@@ -76,6 +76,9 @@ build:
           ./configure --config dev     --projectRoot $WERCKER_SOURCE_DIR --ebEnvName $EB_ENV_NAME --version $(cat $WERCKER_ROOT/VERSION)
           ./configure --config sandbox --projectRoot $WERCKER_SOURCE_DIR --ebEnvName $EB_ENV_NAME --version $(cat $WERCKER_ROOT/VERSION)
     - script:
+        name: check api consistency
+        code: scripts/wercker/run-command ./run exec scripts/api-generator.coffee --check
```
Comment (cihangir) [inline wercker.yml]: 👍

**Review by cihangir — CHANGES_REQUESTED** [review]

**Review by cihangir — APPROVED** [review]
💯

**Pull request merged**